python/cudf/cudf/core/index.py

```diff
@@ -894,22 +892,11 @@ def _binaryop(
             return ret.values
         return ret
 
-    @_cudf_nvtx_annotate
+    # Override just to make mypy happy.
     def _copy_type_metadata(
-        self, other: Frame, include_index: bool = True
+        self: GenericIndex, other: GenericIndex
     ) -> GenericIndex:
```

Comment:
Would the MyPy issues here be resolved if instead of defining `_copy_type_metadata` separately, `BaseIndex` and `Frame` both inherited it from a shared parent class? Or would we just run into the same function signature conflicts between `Frame` and this hypothetical parent class?

Asking because this conflict seems to imply there's some commonality between the two classes that could be consolidated further, though I don't think it's worth it to remove this one-off override.

Reply:
You raise a very good point. The short answer is that no, adding that wouldn't fix this problem, but you are correctly identifying commonalities that we may want to address at some point.

Regarding the type annotation, the issue is the following. Consider the following example:
```
from typing import TypeVar

AT = TypeVar("AT", bound="A")
BT = TypeVar("BT", bound="B")


class A:
    def f(self: AT, other: AT):
        pass


class B:
    def f(self: BT, other: BT):
        pass


class C(A, B):
    pass


class D(B):
    pass
```
Now, imagine that we removed the parameter `other` from both definitions of `f`. In that scenario, this example will pass mypy. mypy can safely infer that `C().f()` will call `C.f(self)` where `self` is an instance of `C`, i.e. it is an instance of a subclass of both `A` and `B`. However, when you have the parameter `other`, mypy has a problem. It thinks `C.f` needs to be compatible with both `A.f` and `B.f`, so `other` could be either a subtype of `A` or of `B`. However, that means that you could call `C.f(D())`, which would violate the signature of `A.f`. Similar issues arise with return types where we need to narrow the return type beyond what mypy would expect to be safe (i.e. return a subclass of both parents, not a subclass of _either_ parent). There may be a more elegant solution for this confusion than what I'm doing in this PR, but it was the most obvious solution that I could come up with.

As to the other point, there's definitely common interfaces between `BaseIndex` and `Frame`. Originally all `Index` classes inherited from `Frame`, but I removed that because it led to all sorts of conceptual issues with `RangeIndex` in particular. We may eventually want to find a way to unify the hierarchy a bit more, but I wouldn't want to do so with anything other than pure abstract classes. Multiple inheritance and diamonds aren't my favorite, it's pretty fragile when the parents aren't all abstract. It's a necessary evil for us, but I'd feel much more comfortable with it if we could minimize the amount of implementation at the highest levels of the hierarchy (and we're definitely working towards that on the Index side of things).